API.md

```diff
@@ -0,0 +1,337 @@
+# API
+
+
+## Asynchrony and idiomatic event driven APIs
+
+Using the API, it's clear that LDK tries to provide an asynchronous, event driven API.
+As example:
+
+> Implementation notes: No methods should be called on ChannelManager until after it has been synced and its chain::Watch has been given the ChannelMonitors as described in the next two steps.
```

Comment:
Right, there's a few answers to this, I think -

1) On the specific issue of `ChannelManager`, I do wonder if we shouldn't just be beefing up `ChannelManagerConstructor` instead - eg create the `ChannelManager` in it, then expose the data/functions necessary to do chain replay, wait for them to be called, then require calling `chain_sync_complete`. We could hide the constructor in `ChannelManager` or just special-case it to have docs that point to `ChannelManagerConstructor` instead.

2) For `Event`s, I think one approach is to do something similar to the persister crate in rust (https://docs.rs/lightning-persister/0.0.13/lightning_persister/struct.FilesystemPersister.html) and have a trait that takes events and then persists the `ChannelManager`. I currently have a draft of this as a part of `ChannelManagerConstructor`, I'll open a PR and ping you on it to see what you think. Its definitely not as fancy as the stream stuff you suggest below.

3) The other immediate API that I think this may be relevant to (but isn't obviously so) is the `Persist` itself around `ChannelMonitor`s. The API is designed to be...either async or not, which definitely makes this complicated. In Rust/the current API, you can *either* return a `Result_NoneChannelMonitorUpdateErrZ.constructor_ok()` for the immediate mode indicating you've persisted the object to disk *or* you can always return `Result_NoneCahnnelMonitorUpdateErrZ.constructor_err(LDKChannelMonitorUpdateErr.LDKChannelMonitorUpdateErr_TemporaryFailure)` (boy does that need to be cleaned up) indicating that the persistence is in-progress and then call `ChannelManager.monitor_update_persisted` later. Do you have a recommendation for a good way to capture the trinary state here - either (a) return immediately, (b) return a future, or (c) return an error.